src/liballoc/collections/btree/node.rs

```diff
@@ -489,33 +528,63 @@ impl<'a, K: 'a, V: 'a, Type> NodeRef<marker::Mut<'a>, K, V, Type> {
-        // More precisely, it is the call to `len` that is the culprit,
-        // because that creates a shared reference to the header, which *can*
-        // overlap with the keys (and even the values, for ZST keys).
-        let len = self.len();
+    /// SAFETY: the caller must ensure that the node has more than `idx` initialized elements.
+    unsafe fn into_key_mut_at(mut self, idx: usize) -> &'a mut K {
+        debug_assert!(idx < self.len());
+
         let leaf = self.as_leaf_mut();
-        // SAFETY: The keys and values of a node must always be initialized up to length.
-        let keys = unsafe {
-            slice::from_raw_parts_mut(MaybeUninit::first_ptr_mut(&mut (*leaf).keys), len)
-        };
-        let vals = unsafe {
-            slice::from_raw_parts_mut(MaybeUninit::first_ptr_mut(&mut (*leaf).vals), len)
-        };
-        (keys, vals)
+        unsafe { leaf.keys.get_unchecked_mut(idx).get_mut() }
+    }
+
+    /// SAFETY: the caller must ensure that the node has more than `idx` initialized elements.
+    unsafe fn into_val_mut_at(mut self, idx: usize) -> &'a mut V {
+        debug_assert!(idx < self.len());
+
+        let leaf = self.as_leaf_mut();
+        unsafe { leaf.vals.get_unchecked_mut(idx).get_mut() }
+    }
+
+    /// SAFETY: the caller must ensure that the node has more than `idx` initialized elements.
+    unsafe fn into_key_mut_val_mut_at(mut self, idx: usize) -> (&'a mut K, &'a mut V) {
+        debug_assert!(idx < self.len());
+
+        // We cannot call into_key_mut_at and into_val_mut_at, because calling the second one
+        // invalidates the reference returned by the first. Even if you `ptr::read(&self)`
+        // to fool the borrow checker, it is still undefined behavior.
+        let leaf = self.as_leaf_mut();
+        let key = unsafe { leaf.keys.get_unchecked_mut(idx).get_mut() };
+        let val = unsafe { leaf.vals.get_unchecked_mut(idx).get_mut() };
+        (key, val)
     }
 }
 
-impl<'a, K, V> NodeRef<marker::Mut<'a>, K, V, marker::Leaf> {
-    /// Adds a key/value pair the end of the node.
+impl<'a, K: 'a, V: 'a, Type> NodeRef<marker::ValMut<'a>, K, V, Type> {
+    /// SAFETY: the caller must ensure that the node has more than `idx` initialized elements.
+    unsafe fn into_key_val_mut_at(self, idx: usize) -> (&'a K, &'a mut V) {
+        debug_assert!(idx < self.len());
+
+        // We only create a reference to the one element we are interested in,
+        // to avoid aliasing with outstanding references to other elements,
+        // in particular, those returned to the caller in earlier iterations.
+        let leaf = self.node.as_ptr();
+        let keys = unsafe { &raw const (*leaf).keys };
+        let vals = unsafe { &raw mut (*leaf).vals };
+        let key0 = keys as *const K;
+        let val0 = vals as *mut V;
+        let key = unsafe { key0.add(idx) };
+        let val = unsafe { val0.add(idx) };
+        unsafe { (&*key, &mut *val) }
```

Comment:
Now that https://github.com/rust-lang/rust/pull/73986 landed you can use those methods, which should make this code a lot simpler. :)
(Be sure to avoid https://github.com/rust-lang/rust/issues/73987 though.)

Reply:
Hm, except that these are arrays of `MaybeUninit`, right? So we need another method to go from `*const/mut MaybeUninit<T>` to `*const/mut T`...

Reply:
I got it in but it doesn't make sense. For instance, before adding type annotation `*mut [_]` onto vals, rustc says "no method named `get_unchecked_mut` found for raw pointer `*mut [core::mem::MaybeUninit<V>; _]` in the current scope". That is the type I expected from a &raw mut. Tell it to be the type it already is, and suddenly rustc says all is well. Am I crazy or is rustc?

Reply:
That sounds odd; do you have a small example demonstrating the problem?

Methods with raw pointer `self` are still fairly new; it is very possible that they have some strange quirks/bugs.

Reply:
That was surprisingly [easy to reproduce](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=fcb75084044ed819f79b19b21d30fecb): omit the type annotation, game over.

Reply:
Ah... we probably do not have implicit unsizing coercions for raw slice pointers.

Reply:
Oh... I didn't know `*const [_]` was more than a shorthand for `*const [i32; 2]` in that context. Would never have found this if I wasn't a lazy typist.

Reply:
It's a shorthand for `*const [i32]` actually.